src/backend/distributed/commands/multi_copy.c

```diff
@@ -218,19 +403,61 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
 		}
 	}
 
-	partitionMethod = PartitionMethod(tableId);
-	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_RANGE)
-	{
-		CopyToExistingShards(copyStatement, completionTag);
-	}
-	else if (partitionMethod == DISTRIBUTE_BY_APPEND)
+	workerCopy = WorkerCopy(copyStatement);
+	if (workerCopy)
 	{
-		CopyToNewShards(copyStatement, completionTag);
+		NodeAddress *masterNodeAddress = MasterNodeAddress(copyStatement);
+		char *nodeName = masterNodeAddress->nodeName;
+		int32 nodePort = masterNodeAddress->nodePort;
+		PGconn *masterConnection = ConnectToNode(nodeName, nodePort);
+
+		char *relationName = copyStatement->relation->relname;
+		PGresult *queryResult = NULL;
+
+		char partitionMethod = RemotePartitionMethod(masterConnection, relationName);
+		if (partitionMethod != DISTRIBUTE_BY_APPEND)
+		{
+			ereport(ERROR, (errmsg("copy from worker nodes are only supported "
+								   "for append-partitioned tables")));
+		}
+
+		/*
+		 * Start a traction to roll back all metadata changes on the master on
+		 * the of a failure.
+		 */
+		queryResult = PQexec(masterConnection, "BEGIN");
+		if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+		{
+			ereport(ERROR, (errmsg("failed to start master node transaction")));
+		}
+
+		RemoveMasterOption(copyStatement);
+		CopyToNewShards(copyStatement, completionTag, masterConnection);
+
+		/* commit metadata transactions */
+		queryResult = PQexec(masterConnection, "COMMIT");
+		if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+		{
+			ereport(ERROR, (errmsg("failed to commit master node transaction")));
```

Comment:
these messages should probably be a bit more descriptive, what does this actually mean to the user?